monitoring/ping.go

```diff
@@ -0,0 +1,308 @@
+		latencyStats:   *latencyTTLMap,
+	}, nil
+}
+
+// pingChecker is a checker that verifies that ping times (RTT) between nodes in
+// the cluster are within a predefined threshold
+type pingChecker struct {
+	self           serf.Member
+	serfClient     serf.RPCClient
+	serfMemberName string
+	latencyStats   ttlmap.TTLMap
+	mux            sync.Mutex
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	// FIXME: #1 RTTThreshold will become configurable in future
+	// FIXME: #2 Send RTTThreshold value to metrics
+
+	err := c.check(ctx, r)
+
+	if err != nil {
+		c.setProbeStatus(ctx, r, err, pb.Probe_Failed)
+		return
+	}
+	c.setProbeStatus(ctx, r, nil, pb.Probe_Running)
+
+	return
+}
+
+// check runs the actual system status verification code and returns an error
+// in case issues arise in the process
+func (c *pingChecker) check(ctx context.Context, r health.Reporter) error {
+
+	client := &c.serfClient
+	if client == nil {
+		return trace.NotFound("serf client not initialized yet")
+	}
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.checkNodesRTT(nodes, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// checkNodesRTT implements the bulk of the logic by checking the ping time
+// between this node (self) and the other Serf Cluster member nodes
+func (c *pingChecker) checkNodesRTT(nodes []serf.Member, client *serf.RPCClient) error {
+	// ping each other node and fail in case the results are over a specified
+	// threshold
+	for _, node := range nodes {
+		// skip pinging self
+		if c.self.Addr.String() == node.Addr.String() {
+			continue
+		}
+
+		rttNanoSec, err := calculateRTT(client, c.self, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = c.saveLatencyStats(rttNanoSec, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		latencyInterface, exists := c.latencyStats.Get(node.Name)
+		if !exists {
+			return trace.NotFound("latency for %s not found", node.Name)
+		}
+		latency, ok := latencyInterface.([]int64)
+		if !ok {
+			return trace.BadParameter("expected latency for %s to be []int64 got %T", c.serfMemberName, latency)
+		}
+		latencyHistogram, err := c.buildLatencyHistogram(node.Name)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		log.Debugf("%s <-ping-> %s = %dns [latest]", c.self.Name, node.Name, rttNanoSec)
+		log.Debugf("%s <-ping-> %s = %dns [%.2f percentile]",
+			c.self.Name, node.Name,
+			latencyHistogram.ValueAtQuantile(latencyQuantile),
+			latencyQuantile)
+
+		latencyPercentile := latencyHistogram.ValueAtQuantile(latencyQuantile)
+		if latencyPercentile >= latencyThreshold.Nanoseconds() {
```

Comment:
I guess that's bound to interpretation. Since it's a percentile I'd be conservative and start alerting as soon as it reaches the threshold... not when it's strictly over the threshold.

But then again it's really bound to interpretation so it can be both ways, I guess? :thinking: :man_shrugging: